### ISSUE TITLE
Stereo tracks & affordances

### ISSUE THREAD

**Issue by jounih** [audacity/audacity#1348]
![stereo clips@2x](https://user-images.githubusercontent.com/10109867/126496025-4c8e4064-03a0-4db9-8260-29ebf9962b24.jpg)

**Comment by petersampsonaudacity** [comment]
With the affordance the upper channel is not aligned properly with the Vertical Scale

**Comment by petersampsonaudacity** [comment]
But it is much better having the single affordance for a stereo track

**Comment by vsverchinsky** [comment]
Do we need to keep resizer between channels, or they should always have same size?

**Comment by petersampsonaudacity** [comment]
> Do we need to keep resizer between channels, or they should always have same size?

Well personally I've never used differently sized channels (I never saw the point - I had no use case where it could/would be useful).

You might need to address this question to a broader audience - the Quality email list and the Forum for example.

**Comment by stmueller** [comment]
> | Do we need to keep resizer between channels, or they should always have same size?

In early versions of audacity, since stereo tracks are basically two mono tracks glued together with some logic, changing the size of one track would resize it but keep the other track unchanged, consequently changing the size of the entire stereo track bundle.  I modified this pre-1.0 to the present behavior that maintains the size of the stereo track when one track is resized, and resizes both proportionally when the stereo track is resized.  I did not do this to handle a specific use case (like wanting to zoom in on one track of a stereo track to use the draw tool), but because it was a simple way to resolve the broken behavior, without overhauling the simple stereo track object.  Disabling this and forcing each track to be 50% of the stereo size is probably reasonable and I'm guessing most other software works like that, but there are probably several bugs that could be introduced if this behavior is changed, especially for cases of merging/splitting tracks mono/stereo tracks.  It is also something that might draw complaints for removing functionality that didn't really negatively impact anything other than the simplicity with which a new feature can be added.  I.e., prepare for bugs filed saying "I can't resize the left stereo track".  It seems the broken thing still is that the affordance eats into the ruler and default track height rather than being added outside the ruler/track height, and forcing 50% track heights for stereo tracks wouldn't do anything for that anyway, other than letting you hard-code 50% in the arithmetic needed to make room for it.

**Comment by Paul-Licameli** [comment]
Sorry, fat fingers there

**Comment by Paul-Licameli** [comment]
Very pleased to have @stmueller here!  I have worked a lot on Audacity for several years and know your name from the comments.

I’m not sure why resizing the channels differently is useful, but in my refactoring of the code in recent years, I conserve old behavior unless I have a good reason not to.

**Comment by Paul-Licameli** [comment]
I added the pinned play head, and later there was a request to make it draggable so it’s not always centered (and I think that is useful for recording, to put it more rightward).  When I did that, I also made double click re-center it.  Maybe that’s a good idea for the stereo channel separator too.

**Comment by Paul-Licameli** [comment]
I think @Tantacrul doesn’t oppose keeping the resizing of stereo channels?  But to reduce the channel separator just to a line.  If that is done, that means the hist testing code needs some work so that the resizing click and drag are still easy to do.  But that’s not a serious difficulty I think.

**Comment by Tantacrul** [comment]
> With the affordance the upper channel is not aligned properly with the Vertical Scale (the lower channel, now with no affordance is properly aligned with the Vertical Scale).
> 
> ![image](https://user-images.githubusercontent.com/50205984/126528607-2105cf3d-6340-434e-a67a-2067e23dde9a.png)
> 
> This is presumably because the affordance is eating some of what was formerly the upper channel - rather than being additional screen real-estate.

I imagine it's a mistake in the mockup (@jounih ?). The height of L & R should surely be the same (although L will have extra vertical for the affordance)

**Comment by Tantacrul** [comment]
> I think @Tantacrul doesn’t oppose keeping the resizing of stereo channels? But to reduce the channel separator just to a line. If that is done, that means the hist testing code needs some work so that the resizing click and drag are still easy to do. But that’s not a serious difficulty I think.

Yeah, that's pretty much what I think. Two affordances is obviously a little strange and I think we can keep the resizing hit area the same, while making the vertical distance between waves visibly shorter. This would not be the case when the user joins mono tracks, which I think would need to stay the same as is.

**Comment by Paul-Licameli** [comment]
I think the implementation need not remember that a stereo track was made from mono tracks, but rather could check whether clip boundaries in opposite tracks correspond, and that this check would be cheap enough in the usual case.

**Comment by jounih** [comment]
> With the affordance the upper channel is not aligned properly with the Vertical Scale (the lower channel, now with no affordance is properly aligned with the Vertical Scale).
> 
> ![image](https://user-images.githubusercontent.com/50205984/126528607-2105cf3d-6340-434e-a67a-2067e23dde9a.png)
> 
> This is presumably because the affordance is eating some of what was formerly the upper channel - rather than being additional screen real-estate.

Fixed - good eyes! 

The vertical scale should stop where the affordance starts.

**Comment by petersampsonaudacity** [comment]
> 
> 
> I think the implementation need not remember that a stereo track was made from mono tracks, but rather could check 
>whether clip boundaries in opposite tracks correspond, and that this check would be cheap enough in the usual case.

But clip boundaries in the two channels may not necessarily correspond.

My use case would be 

1) taking say track 1 from Sgt Pepper's Lonely Hearts Club Band (stereo version)
2) split to mono
3) create clips for various sections in each split mono track
4) observe they don't correspond (the L & R channels are considerably different in parts)

**Comment by vsverchinsky** [comment]
> But clip boundaries in the two channels may not necessarily correspond.
In that case each clip should have it's own affordance (like they do have them now in the latest master), despite number of channels in track, right @jounih?

**Comment by petersampsonaudacity** [comment]
And when I remake those mono tracks back into stereo - this is what I get  :-)

![image](https://user-images.githubusercontent.com/50205984/126639913-ae0da9b0-2f66-4678-9b8c-82cc66337ac7.png)

**Comment by petersampsonaudacity** [comment]
> In that case each clip should have it's own affordance (like they do have them now in the latest master), despite number of channels in track, right @jounih?

Probably yes - but it is possibly an unusual use case.

One might argue _(and I probably would)_  that in turning it back into stereo that the mono clips should propagate across both channels - so you get stereo clips throughout.

**Comment by jounih** [comment]
> > But clip boundaries in the two channels may not necessarily correspond.
> 
> In that case each clip should have it's own affordance (like they do have them now in the latest master), despite number of channels in track, right @jounih?

Yes that's correct - we discussed this with @Tantacrul and @Paul-Licameli yesterday and agreed that in the case of joining mono tracks to stereo, they should retain their individual affordances - exactly for the reason you mentioned. It's an usual case but could happen

**Comment by AnitaBats** [comment]
done (this issue was used as Epic, multiple tasks were created)
